src/serializer/visitors.js

```diff
@@ -150,7 +146,34 @@ export let ClosureRefReplacer = {
       // Un-interpreted nested function.
       return;
     }
-    replaceNestedFunction(functionTag, path, state);
+    const duplicateFunctionInfo = state.factoryFunctionInfos.get(functionTag);
+    if (duplicateFunctionInfo && canShareFunctionBody(duplicateFunctionInfo)) {
+      const { factoryId } = duplicateFunctionInfo;
+      path.replaceWith(t.callExpression(t.memberExpression(factoryId, t.identifier("bind")), [nullExpression]));
+    }
+  },
+
+  // Replace "function foo() {}" ==> "var foo = factory_id.bind(null)".
```

Comment:
At runtime, function declarations get hoisted, but declaration initializers are not!

Try a test case like this:
```js
f = function() {
  return nested; // note that declaration comes later --- this is okay!
  function nested() {
    var x0 = 1;
    var x1 = x0 + x0;
    var x2 = x1 + x1;
    var x3 = x2 + x2;
    var x4 = x3 + x3;
    return x4;
  }
}
g = f();
inspect = function() {  
  return g();
}
```

Reply:
Wow, good insight, thanks for catching this!
I will use this PR only for non-factory function case, and use a part3 PR for the function declaration.